### PR TITLE
[Feature] $filter path segment

### PR DESCRIPTION
**Pull request by AlanWong-MS**
### Description
Implementation of the $filter path segment as part of the [set-based operation](http://docs.oasis-open.org/odata/new-in-odata/v4.01/cn01/new-in-odata-v4.01-cn01.html#_Toc485385072) in the OData v4.01 specification ([ANBF](http://docs.oasis-open.org/odata/odata/v4.01/csprd04/abnf/)).

This PR includes the $filter path segment feature, which includes a new class FilterSegment, modifications to the URI parser to process the new path segment, unit test cases for the class itself, and functional test cases for the use of the filter path segment.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary
See Checklist.

### CODE REVIEW

File: test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/SemanticAst/FilterSegmentTests.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            FilterSegment filterSegment1 = new FilterSegment(alias, filterClause.RangeVariable, HardCodedTestModel.GetPet1Type(), false);
+            FilterSegment filterSegment2 = new FilterSegment(alias, filterClause.RangeVariable, HardCodedTestModel.GetPersonType(), false);
+
+            filterSegment1.Equals(filterSegment2).Should().BeFalse();
+            filterSegment2.Equals(filterSegment1).Should().BeFalse();
+        }
+        #endregion
+
+        #region Helper Functions
+
```
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/SemanticAst/FilterSegmentTests.cs]: nit: empty line

File: src/Microsoft.OData.Core/UriParser/SemanticAst/FilterSegment.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            ExceptionUtils.CheckArgumentNotNull(alias, "alias");
+            ExceptionUtils.CheckArgumentNotNull(rangeVariable, "rangeVariable");
+            ExceptionUtils.CheckArgumentNotNull(bindingType, "bindingType");
+
+            this.alias = alias;
+            this.rangeVariable = rangeVariable;
+            this.bindingType = bindingType;
+
+            this.Identifier = UriQueryConstants.FilterSegment;
+            this.SingleResult = singleResult; // !!! Check during review
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/SemanticAst/FilterSegment.cs]: Verify these during review.

File: src/Microsoft.OData.Client/XmlConstants.cs [hunk 3]
```diff
@@ -300,6 +300,9 @@ internal static class XmlConstants
         /// <summary>A segment name in a URI that indicates that this is a count operation.</summary>
         internal const string UriCountSegment = "$count";
 
+        /// <summary>A segment name in a URI that indicates that this is a filter operation.</summary>
+        internal const string UriFilterSegment = "$filter";
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Client/XmlConstants.cs]: Move this below the CountTrueOption.

File: src/Microsoft.OData.Core/Microsoft.OData.Core.cs [hunk 4]
```diff
@@ -764,9 +764,11 @@ internal sealed class TextRes {
         internal const string CustomUriFunctions_AddCustomUriFunction_CustomFunctionOverloadExists = "CustomUriFunctions_AddCustomUriFunction_CustomFunctionOverloadExists";
         internal const string RequestUriProcessor_InvalidValueForEntitySegment = "RequestUriProcessor_InvalidValueForEntitySegment";
         internal const string RequestUriProcessor_InvalidValueForKeySegment = "RequestUriProcessor_InvalidValueForKeySegment";
+        internal const string RequestUriProcessor_NoNavigationSourceFoundForFilter = "RequestUriProcessor_NoNavigationSourceFoundForFilter";
+        internal const string RequestUriProcessor_FilterPathSegmentRequiresParameterAlias = "RequestUriProcessor_FilterPathSegmentRequiresParameterAlias";
         internal const string RequestUriProcessor_EmptySegmentInRequestUrl = "RequestUriProcessor_EmptySegmentInRequestUrl";
         internal const string RequestUriProcessor_SyntaxError = "RequestUriProcessor_SyntaxError";
-        internal const string RequestUriProcessor_CountOnRoot = "RequestUriProcessor_CountOnRoot";
+        internal const string RequestUriProcessor_NotAllowedOnRoot = "RequestUriProcessor_NotAllowedOnRoot";
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/Microsoft.OData.Core.cs]: For good practices, don't modify existing strings/string names for loc files.

File: src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs [hunk 5]
```diff
@@ -449,17 +449,145 @@ private bool TryCreateCountSegment(string segmentText)
...
+            /*
+             * 1) Check whether the path segment starts with $filter.
+             * 2) Ensure that "=@" that follows the identifier and that there's only one '@'.
+             * 3) Extract the parameter alias and validate it syntactically.
+             * 4) Add the filter segment to list of parsed segments.
+             */
+
+            // 1) Check whether the path segment starts with $filter. Past this point, we will throw invalid syntax exceptions
+            // because we will assume that the user is attempting to use the $filter path segment.
+            if (!segmentText.StartsWith(UriQueryConstants.FilterSegment, StringComparison.OrdinalIgnoreCase))
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs]: Use EnableCaseInsensitiveUriFunctionIdentifier (which needs to be renamed to remove "UriFunctionIdentifier").

File: src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs [hunk 6]
```diff
@@ -449,17 +449,145 @@ private bool TryCreateCountSegment(string segmentText)
...
+            {
+                return false;
+            }
+
+            int index = UriQueryConstants.FilterSegment.Length;
+            if (index == segmentText.Length ||
+                segmentText.Substring(index, JsonLightConstants.FunctionParameterAssignment.Length) != JsonLightConstants.FunctionParameterAssignment ||
+                segmentText.LastIndexOf('@') != index + 1)
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs]: Add another check that ensures that we have something after @.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/ParameterAliasFunctionalTests.cs [hunk 7]
```diff
@@ -335,6 +334,452 @@ public void ParseFilter_AliasInFunction_BuiltIn()
...
+                    aliasNodes["@p1"].ShouldBeBinaryOperatorNode(BinaryOperatorKind.Equal).And.Right.ShouldBeConstantQueryNode("num");
+                    aliasNodes["@p2"].ShouldBeBinaryOperatorNode(BinaryOperatorKind.Equal).And.Right.ShouldBeConstantQueryNode(1);
+                });
+        }
+
+        [Fact]
+        public void ParseFilter_AliasInFilterPathSegment_NavigationProperty()
+        {
+            ParseUriAndVerify(
+                new Uri("http://gobbledygook/People/1/MyDog/$filter=@p1?@p1=true"),
```
Comment (AlanWong-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/ParameterAliasFunctionalTests.cs]: Parameter alias can't be applied on singletons.

File: src/Microsoft.OData.Core/UriParser/SemanticAst/FilterSegment.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+        /// <summary>
+        /// Build a segment representing $filter.
+        /// </summary>
+        /// <param name="alias">Corresponding parameter alias for the $filter path segment.</param>
+        /// <param name="rangeVariable">An expression that represents a single value from the collection.</param>
+        /// <param name="bindingType">The type that this segment is bound to.</param>
+        /// <param name="singleResult">Whether the segment targets a single result.</param>
+        /// <exception cref="System.ArgumentNullException">Throws if any input parameter is null.</exception>
+        public FilterSegment(ParameterAliasNode alias, RangeVariable rangeVariable, IEdmType bindingType, bool singleResult)
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/SemanticAst/FilterSegment.cs]: Require navigationSource

File: src/Microsoft.OData.Core/Microsoft.OData.Core.txt [hunk 9]
```diff
@@ -841,10 +841,13 @@ CustomUriFunctions_AddCustomUriFunction_CustomFunctionOverloadExists=The given f
 RequestUriProcessor_InvalidValueForEntitySegment=The ODataPathSegment provided (Id = {0}) is not an EntitySetSegment.
 RequestUriProcessor_InvalidValueForKeySegment=The KeySegment provided (Id = {0}) is either null, having no keys, or does not target a single resource.
 
+RequestUriProcessor_NoNavigationSourceFoundForFilter=There are no navigation sources found to apply $filter.
+RequestUriProcessor_FilterPathSegmentRequiresParameterAlias=The $filter path segment must contain a single parameter alias. In other words, it must have the form '$filter=@...', where the alias name goes in the ellipsis.
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/Microsoft.OData.Core.txt]: Reword exception message

File: src/Microsoft.OData.Core/UriParser/Visitors/PathSegmentToResourcePathTranslator.cs [hunk 10]
```diff
@@ -176,6 +176,17 @@ public override string Translate(CountSegment segment)
...
 
+        /// <summary>
+        /// Translate a FilterSegment
+        /// </summary>
+        /// <param name="segment">the segment to Translate</param>
+        /// <returns>Defined by the implementer.</returns>
+        public override string Translate(FilterSegment segment)
+        {
+            Debug.Assert(segment != null, "segment != null");
+            return "/" + segment.Identifier;
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Visitors/PathSegmentToResourcePathTranslator.cs]: Is this correct?

File: src/Microsoft.OData.Core/UriParser/Visitors/PathSegmentToContextUrlPathTranslator.cs [hunk 11]
```diff
@@ -169,6 +169,16 @@ public override string Translate(CountSegment segment)
...
         }
 
+        /// <summary>
+        /// Translate a FilterSegment
+        /// </summary>
+        /// <param name="segment">the segment to Translate</param>
+        /// <returns>Defined by the implementer.</returns>
+        public override string Translate(FilterSegment segment)
+        {
+            return string.Empty;
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Visitors/PathSegmentToContextUrlPathTranslator.cs]: Correct?

File: src/Microsoft.OData.Core/UriParser/Visitors/IsCollectionTranslator.cs [hunk 12]
```diff
@@ -86,6 +86,18 @@ public override bool Translate(CountSegment segment)
...
+        /// <summary>
+        /// Translate a FilterSegment
+        /// </summary>
+        /// <param name="segment">the segment to Translate</param>
+        /// <returns>UserDefinedValue</returns>
+        /// <exception cref="System.ArgumentNullException">Throws if the input segment is null.</exception>
+        public override bool Translate(FilterSegment segment)
+        {
+            ExceptionUtils.CheckArgumentNotNull(segment, "segment");
+            return false;
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Visitors/IsCollectionTranslator.cs]: Return true. #done

File: src/Microsoft.OData.Core/UriParser/SemanticAst/DetermineNavigationSourceTranslator.cs [hunk 13]
```diff
@@ -138,6 +138,18 @@ public override IEdmNavigationSource Translate(CountSegment segment)
...
+        /// <summary>
+        /// Determine the NavigationSource of a FilterSegment
+        /// </summary>
+        /// <param name="segment">The FilterSegment to look in.</param>
+        /// <returns>null, since $filter doesn't have an navigation source</returns>
+        /// <exception cref="System.ArgumentNullException">Throws if the input segment is null.</exception>
+        public override IEdmNavigationSource Translate(FilterSegment segment)
+        {
+            ExceptionUtils.CheckArgumentNotNull(segment, "segment");
+            return null;
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/SemanticAst/DetermineNavigationSourceTranslator.cs]: Return navigationSource. #done

File: src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs [hunk 14]
```diff
@@ -516,9 +644,9 @@ private bool TryCreateEntityReferenceSegment(string text)
             this.parsedSegments.Add(navPropLinkSegment);
 
             // Add key segments back to the parsed segments
-            while (keySegmentsForPreviousPathSegment.Count > 0)
+            while (keyAndFilterSegmentsForPreviousPathSegment.Count > 0)
             {
-                this.parsedSegments.Add(keySegmentsForPreviousPathSegment.Pop());
+                this.parsedSegments.Add(keyAndFilterSegmentsForPreviousPathSegment.Pop());
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs]: Some of the assumptions above are not correct. To preserve backwards compatibility, revert the changes for the $ref and add a new ReferenceSegment to add corrected behavior.

File: src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs [hunk 15]
```diff
@@ -449,17 +449,145 @@ private bool TryCreateCountSegment(string segmentText)
             return true;
         }
 
+        /// <summary>
+        /// Retrieves the last navigation source and index of the navigation source in the parsed segments.
+        /// </summary>
+        /// <param name="navigationSource">Extracted navigation source. This is null if no source is found.</param>
+        /// <param name="navigationSourceIndex">Extracted index of the navigation source. This is -1 if no source is found.</param>
+        /// <returns>True if a navigation sourse was found; false, otherwise.</returns>
+        private bool GetLastNavigationSource(out IEdmNavigationSource navigationSource, out int navigationSourceIndex)
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs]: Can we assume that the previous segment has the navigationSource info?

Comment (AlanWong-MS) [reply src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs]: OperationSegments currently do not necessarily have navigationSources, so we cannot assume the previous segment to have navigationSource info. Going to remove this method and instead add a private member to store the last parsed navigationSource.

File: src/Microsoft.OData.Core/Microsoft.OData.Core.txt [hunk 16]
```diff
@@ -842,10 +842,15 @@ CustomUriFunctions_AddCustomUriFunction_CustomFunctionOverloadExists=The given f
 RequestUriProcessor_InvalidValueForEntitySegment=The ODataPathSegment provided (Id = {0}) is not an EntitySetSegment.
 RequestUriProcessor_InvalidValueForKeySegment=The KeySegment provided (Id = {0}) is either null, having no keys, or does not target a single resource.
 
+RequestUriProcessor_CannotApplyFilterOnSingleEntities=$filter path segment cannot be applied on single entities or singletons. Entity type: '{0}'.
+RequestUriProcessor_NoNavigationSourceFound=There are no navigation sources found to apply '{0}'.
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Microsoft.OData.Core.txt]: >RequestUriProcessor_NoNavigationSourceFound [](start = 0, length = 43)

...._NavigationSourceNotFound

File: src/Microsoft.OData.Core/Microsoft.OData.Core.txt [hunk 17]
```diff
@@ -842,10 +842,15 @@ CustomUriFunctions_AddCustomUriFunction_CustomFunctionOverloadExists=The given f
 RequestUriProcessor_InvalidValueForEntitySegment=The ODataPathSegment provided (Id = {0}) is not an EntitySetSegment.
 RequestUriProcessor_InvalidValueForKeySegment=The KeySegment provided (Id = {0}) is either null, having no keys, or does not target a single resource.
 
+RequestUriProcessor_CannotApplyFilterOnSingleEntities=$filter path segment cannot be applied on single entities or singletons. Entity type: '{0}'.
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Microsoft.OData.Core.txt]: >$filter [](start = 54, length = 7)

The $filter

File: src/Microsoft.OData.Core/Microsoft.OData.Core.txt [hunk 18]
```diff
@@ -842,10 +842,15 @@ CustomUriFunctions_AddCustomUriFunction_CustomFunctionOverloadExists=The given f
...
 
+RequestUriProcessor_CannotApplyFilterOnSingleEntities=$filter path segment cannot be applied on single entities or singletons. Entity type: '{0}'.
+RequestUriProcessor_NoNavigationSourceFound=There are no navigation sources found to apply '{0}'.
+RequestUriProcessor_FilterPathSegmentRequiresParameterAlias=The $filter path segment must contain a single parameter alias in the form $filter=@alias.
+
 ; Copied from the server, will be incrementally replaced.
 RequestUriProcessor_EmptySegmentInRequestUrl=Empty segment encountered in request URL. Please make sure that a valid request URL is specified.
 RequestUriProcessor_SyntaxError=Bad Request - Error in query syntax.
 RequestUriProcessor_CountOnRoot=The request URI is not valid, the segment $count cannot be applied to the root of the service.
+RequestUriProcessor_FilterOnRoot=The request URI is not valid, the segment $filter cannot be applied to the root of the service.
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/Microsoft.OData.Core.txt]: >segment $filter [](start = 67, length = 15)

..., the $filter path segment ...

File: src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs [hunk 19]
```diff
@@ -893,15 +1022,23 @@ private bool TryCreateSegmentForOperation(ODataPathSegment previousSegment, stri
...
+             * Try to match an OData segment(starting with �$�).
+             *   - Note: $filter path segment is a special case that has the format "$filter=@a", where @a represents an alias.
+             * Try to match an alias - qualified bound action name, bound function overload, or type name.
+             * Try to match a namespace-qualified bound action name, bound function overload, or type name.
+             * Try to match an unqualified bound action name, bound function overload, or type name in a default namespace.
+             * Treat as a key.
+             */
+
             // $value
-            if (this.TryCreateValueSegment(text))
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs]: >TryCreateValueSegment [](start = 21, length = 21)

maybe keep it unchanged to use "text"?

File: src/Microsoft.OData.Core/UriParser/Visitors/PathSegmentToResourcePathTranslator.cs (hunk 10 above)
Comment (xuzhg) [reply src/Microsoft.OData.Core/UriParser/Visitors/PathSegmentToResourcePathTranslator.cs]: why do you mention to modifi to return "$filter"? It's a referenceSegment.

---
In reply to: [170140762](https://github.com/OData/odata.net/pull/1069#discussion_r170140762) [](ancestors = 170140762)

File: src/Microsoft.OData.Edm/Schema/CoreModel/EdmCoreModelComplexType.cs [hunk 20]
```diff
@@ -20,68 +20,95 @@ internal sealed class EdmCoreModelComplexType : EdmType, IEdmComplexType, IEdmCo
...
-        public override EdmTypeKind TypeKind => EdmTypeKind.Complex;
+        public override EdmTypeKind TypeKind
+        {
+            get { return EdmTypeKind.Complex; }
+        }
 
         /// <summary>
         /// Gets the schema element kind of this type.
         /// </summary>
-        public EdmSchemaElementKind SchemaElementKind => EdmSchemaElementKind.TypeDefinition;
```
Comment (xuzhg) [inline src/Microsoft.OData.Edm/Schema/CoreModel/EdmCoreModelComplexType.cs]: >=> [](start = 54, length = 2)

something wrong with this syntax?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/ParameterAliasFunctionalTests.cs [hunk 21]
```diff
@@ -335,6 +334,494 @@ public void ParseFilter_AliasInFunction_BuiltIn()
...
+                {
+                });
+            parse.ShouldThrow<ODataException>().WithMessage(ODataErrorStrings.RequestUriProcessor_ResourceNotFound("$filter=@p1&@p1=true"));
+        }
+
+        [Fact]
+        public void ParseFilter_AliasInFilterPathSegment_FilterSegmentOnSingleton()
+        {
+            Action parse = () => ParseUriAndVerify(
+                new Uri("http://gobbledygook/Boss/$filter=@p1?@p1=true"),
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/ParameterAliasFunctionalTests.cs]: >$filter=@p1?@p1=true [](start = 50, length = 20)

Would you please make sure $filter can follow up the singleton? What's the scenario?

**Comment by xuzhg** [comment]
I approved with some comments to unblock you. I missed the design meeting, but the structure looks good to me.

File: src/Microsoft.OData.Core/Microsoft.OData.Core.txt (hunk 17 above)
Comment (AlanWong-MS) [reply src/Microsoft.OData.Core/Microsoft.OData.Core.txt]: Done.

File: src/Microsoft.OData.Core/Microsoft.OData.Core.txt (hunk 16 above)
Comment (AlanWong-MS) [reply src/Microsoft.OData.Core/Microsoft.OData.Core.txt]: Done.

File: src/Microsoft.OData.Core/Microsoft.OData.Core.txt (hunk 18 above)
Comment (AlanWong-MS) [reply src/Microsoft.OData.Core/Microsoft.OData.Core.txt]: I am going to keep as-is to make it consistent with the RequestUriProcessor_CountOnRoot message. I think it's clear enough that it's a path segment.

File: src/Microsoft.OData.Edm/Schema/CoreModel/EdmCoreModelComplexType.cs (hunk 20 above)
Comment (AlanWong-MS) [reply src/Microsoft.OData.Edm/Schema/CoreModel/EdmCoreModelComplexType.cs]: The code was fine, but Stylecop was giving warnings about it and I was not able to suppress the specific warning. So I updated the syntax to get rid of the Stylecop warning.

File: src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs (hunk 19 above)
Comment (AlanWong-MS) [reply src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs]: We had a discussion during a design meeting (I think you were away at the time). I updated this to use `identifier` and `parenthesisExpression` because in the previous implementation, each of the segment parsers were calling `ExtractSegmentIdentifierAndParenthesisExpression` which was redundant work. I moved the `ExtractSegmentIdentifierAndParenthesisExpression` out of those methods and to line 1027 and then I pass the information into the `TryCreate...` methods.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/ParameterAliasFunctionalTests.cs (hunk 21 above)
Comment (AlanWong-MS) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/ParameterAliasFunctionalTests.cs]: What is the scenario for $filter following singleton? I think it's correct that $filter cannot follow singleton because there is only one entity and filtering should apply on a collection.

**Comment by AlanWong-MS** [comment]
Closing this in favor of #1110, which is a superset of this change.
